integration-libs/epd-visualization/components/visual-viewer/visual-viewer.service.ts

```diff
@@ -0,0 +1,1496 @@
+import { VisualizationLookupService } from '@spartacus/epd-visualization/root';
+
+// @ts-ignore
+import * as ui5 from '@sapui5/ts-types-esm';
+
+import type Core from 'sap/ui/core/Core';
+import type { CSSColor } from 'sap/ui/core/library';
+
+import type Scene from 'sap/ui/vk/Scene';
+import type Viewport from 'sap/ui/vk/Viewport';
+import type ContentConnector from 'sap/ui/vk/ContentConnector';
+import type AnimationPlayer from 'sap/ui/vk/AnimationPlayer';
+import type DrawerToolbar from 'sap/ui/vk/DrawerToolbar';
+import type ViewStateManager from 'sap/ui/vk/ViewStateManager';
+type ViewManager = any;
+type NodeRef = any;
+import type UIArea from 'sap/ui/core/UIArea';
+import type ContentResource from 'sap/ui/vk/ContentResource';
+
+import { NavigationMode } from './models/navigation-mode';
+import { SelectionMode } from './models/selection-mode';
+import { NodeContentType } from './models/node-content-type';
+import { SceneLoadState } from './models/scene-load-state';
+import { LoadedSceneInfo, SceneLoadInfo } from './models/scene-load-info';
+import { ContentType } from '@spartacus/epd-visualization/root';
+import NodeHierarchy from 'sap/ui/vk/NodeHierarchy';
+import { catchError, filter, mergeMap, shareReplay } from 'rxjs/operators';
+import { VisualizationInfo } from '@spartacus/epd-visualization/root';
+import {
+  VisualizationLoadInfo,
+  VisualizationLoadStatus,
+  VisualizationLookupResult,
+} from './models/visualization-load-info';
+import { ZoomTo } from './models/zoom-to';
+
+interface VisualContentChangesFinishedEvent {
+  content: any;
+  failureReason: any;
+}
+
+interface VisualContentLoadFinishedEvent {}
+
+@Injectable({
+  providedIn: 'any',
+})
+export class VisualViewerService {
+  constructor(
+    protected epdVisualizationConfig: EpdVisualizationConfig,
+    protected _sceneNodeToProductLookupService: SceneNodeToProductLookupService,
+    protected visualizationLookupService: VisualizationLookupService,
+    protected elementRef: ElementRef,
+    protected changeDetectorRef: ChangeDetectorRef
+  ) {
+    const ui5BootStrapped$: Observable<void> =
+      this.bootstrapUi5('ui5bootstrap');
+    const ui5Initialized$: Observable<void> = ui5BootStrapped$.pipe(
+      mergeMap(this.initializeUi5.bind(this))
+    );
+    this.viewportAdded$ = ui5Initialized$.pipe(
+      mergeMap(this.addViewport.bind(this)),
+      shareReplay()
+    );
+    this.executeWhenSceneLoaded(this.setInitialPropertyValues.bind(this));
+  }
+
+  protected _ui5: ui5;
+
+  private get sceneNodeToProductLookupService(): SceneNodeToProductLookupService {
+    return this._sceneNodeToProductLookupService;
+  }
+  private set sceneNodeToProductLookupService(
+    value: SceneNodeToProductLookupService
+  ) {
+    this._sceneNodeToProductLookupService = value;
+  }
+
+  private _scene: Scene;
+  private get scene(): Scene {
+    return this._scene;
+  }
+  private set scene(value: Scene) {
+    this._scene = value;
+  }
```

Comment:
This service (and few others like `visual-picking-product-list.service`) contains many groups of: `private` value/setter/getter. What's the advantage of it, since they all private? Did you consider using simply the property value?

Reply:
I am only doing this in order to have getter/setter functions to spy on in unit test code.
I didn't find a way to intercept access to a plain property.

Reply:
IMHO the ability to test something conveniently should not be the reason to shape differently the implementation (especially not to decrease readability).
The getters/setters are not justified from the implementation point of view in this case. So let's look for alternative ways of testing.

Did you consider simply assigning values in the tests, instead of spying on getters/setters? e.g. `visualViewerService['viewport'] = mockViewport`

I understand that then we might not be able to check if the private `viewport` property was accessed during the test (in other words if the `viewport`'s getter spy was invoked once). But such assertion could be considered as too detailed and too much implementation-specific

Unit tests generally shouldn't assert on implementation details, e.g. "what internal properties were accessed", but rather test external output and external side effects when given specific inputs and circumstances. Of course there might be exceptions to this rule, when justified.

I hope you find alternative way that avoids "artificial" getters/setters in the implementation and provide sufficient test scenarios coverage. :) 

That said, I'm really opened for your point of view. What do you think?

Reply:
I would like to be able to do this, but I don't think it's going to be feasible in terms of the time constraints we have.

Reply:
Fair enough. Agreed, it can be postponed - since those are private properties, the refactoring to simple properties shouldn't bring breaking changes.
Do you think you could prioritize it? If yes, when do you think it would be possible?